### PR TITLE
Add PSR-17 `StreamFactoryInterface` Support in `ImplicitHeadMiddleware`

### LINKED ISSUE #29: Add PSR-17 support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | yes
| BC Break    | no

#### Summary

Provide support for using [PSR-17](https://www.php-fig.org/psr/psr-17/) stream factory instead of fetching `StreamInterface` as a service. See mezzio/mezzio#81 for details.

### PR DESCRIPTION
**Pull request by gsteel**
|    Q          |   A
|-------------- | ------
| BC Break      | yes
| New Feature   | yes

### Description

Replaces all callable parameters and properties with `StreamFactoryInterface` and `ResponseFactoryInterface` where appropriate and removes the callable compatability layer.

Closes #29

### CODE REVIEW

**Comment by froschdesign** [comment]
@gsteel 
The target branch 4.0.x is missing.

**Comment by gsteel** [comment]
> The target branch 4.0.x is missing.

I'll create the branch if this is approved 👍

**Comment by boesing** [comment]
Can we please go with the logic as in https://github.com/mezzio/mezzio-authorization/pull/11/files?
Forward compatibility in current major versions should be the goal to reduce the amount of pain when migrating to new majors.
We can fix our internal types in the next major version but until then, providing support for both `callable` and the factory interfaces is doable. There are some caveats (i.e. due to the `aliases` / `delegators` registered in upstream projects - also `initializers` but I forgot these in the other components already) but IMHO worth doing so.

**Comment by gsteel** [comment]
I'm not sure I understand the point of preserving support for callables @boesing - As this lib is only useful to a Mezzio app, it's most likely that Stream|Response factories are already wired up in the container via Diactoros, making migration a zero effort chore.

The callable response decorator was already deprecated, and I'll have to introduce a callable stream decorator to satisfy `StreamFactoryInterface` which in turn will create a hard dependency on Diactoros' `StreamFactory` so that `createStreamFromFile` etc can be implemented.

I thought that this patch was more about removing support for callables than adding PSR-17 because at least `ResponseFactoryInterface` has been supported for quite a while now???

**Comment by boesing** [comment]
> I thought that this patch was more about removing support for callables than adding PSR-17 because at least ResponseFactoryInterface has been supported for quite a while now???

It is not supported in this component for quite a while and thus I dont understand why you think it is already supported.
The whole point of introducing PSR-17 is that we provide the possibility from within all the components before we require upstream to provide PSR-17.
Just because you have `laminas/laminas-diactoros` required does not mean that you automatically have its config provider registered (which would be required to have the factory interfaces available via `$container` - or upstream registered manually all the factory interfaces which is most likely not the case as it is not required to do so in mezzio v3). 
Furthermore, diactoros is not required in mezzio, you can use `nyholm/psr7` which ~(AFAIR) does not provide any PSR-17 factories~ provides PSR-17 factories but would require configuration changes to actually register all the factory interfaces to the `nyholm/psr7` PSR-17 factory).

So overall, adding support for both makes sense as a first step while dropping support in next major.
Starting with mezzio v4, I'd expect all components do support PSR-17 as there won't be any callable registered anymore.

But maybe I am missing something, can you probably elaborate why you think that there is already PSR-17 support in this component?


----

So all I want to say is, I spent quite a lot of time in providing proper forward compatibility for PSR-17 so that we can switch to full PSR-17 in the next major. All I want is that we provide compatibility layer in `mezzio-router` v3 so that we can do this change (so the change in this PR) in `mezzio-router` v4.

**Comment by gsteel** [comment]
@boesing - I reverted the patch and added compatibility for `StreamFactoryInterface` - I was actually targeting 4 here, so I think I understand now that whilst the response factory was already supported here, the stream factory was still outstanding for the 3.x series. I will revert a couple more things here so that this can go into 3.x

File: src/Middleware/ImplicitHeadMiddleware.php [hunk 1]
```diff
@@ -44,19 +47,25 @@ class ImplicitHeadMiddleware implements MiddlewareInterface
 {
     public const FORWARDED_HTTP_METHOD_ATTRIBUTE = 'forwarded_http_method';
 
-    /** @var callable(): StreamInterface */
+    /** @var callable(): StreamInterface|StreamFactoryInterface */
     private $streamFactory;
 
     /**
-     * @param callable(): StreamInterface $streamFactory A factory capable of returning an empty
+     * @param callable(): StreamInterface|StreamFactoryInterface $streamFactory A factory capable of returning an empty
```
Comment (boesing) [inline src/Middleware/ImplicitHeadMiddleware.php]: I guess you have to surround the callable with brackets as psalm might read it as the return value of the callable.

`callable(): StreamInterface|StreamFactoryInterface` !== `(callable():StreamInterface)|StreamFactoryInterface`

I wonder why psalm is happy with the current notation.

File: src/Middleware/ImplicitHeadMiddleware.php [hunk 2]
```diff
@@ -44,19 +47,25 @@ class ImplicitHeadMiddleware implements MiddlewareInterface
...
-    public function __construct(private RouterInterface $router, callable $streamFactory)
+    public function __construct(private RouterInterface $router, callable|StreamFactoryInterface $streamFactory)
     {
-        // Factory is wrapped in closure in order to enforce return type safety.
-        $this->streamFactory = function () use ($streamFactory): StreamInterface {
-            return $streamFactory();
-        };
+        if (is_callable($streamFactory)) {
+            // Factory is wrapped in closure in order to enforce return type safety.
+            $this->streamFactory = function () use ($streamFactory): StreamInterface {
```
Comment (boesing) [inline src/Middleware/ImplicitHeadMiddleware.php]: I actually removed the callable functionality from within the other classes which consumed callables.

Can we rewrite this so that the `streamFactory` property only accepts `StreamFactoryInterface`? That would require some `CallableStreamFactory` which implements the `StreamFactoryInterface` of course.

File: src/Middleware/ImplicitHeadMiddlewareFactory.php [hunk 3]
```diff
@@ -36,6 +36,14 @@ public function __invoke(ContainerInterface $container): ImplicitHeadMiddleware
             );
         }
 
+        // Prefer StreamFactoryInterface if present over StreamInterface callable
```
Comment (boesing) [inline src/Middleware/ImplicitHeadMiddlewareFactory.php]: I am not sure if we should prefer it by default. Thats why I added that trait in the other components (and maybe here?) to detect whether or not there are projects implementing custom logic to provide the `StreamInterface`.

Not sure if I can properly phrase the problem, so if you can't follow me here, just let me know.

**Review by boesing — CHANGES_REQUESTED** [review]
Works for me, the only thing which might happen is that there are custom factories provided by upstream factories for the `StreamInterface` which could be slightly modified to i.e. decorate the `StreamInterface` by default which won't be the case once the `StreamFactoryInterface` creates the `StreamInterface`.

If you need more input on the problem I try to outline, just let me know.
I guess you can adapt the tests regarding the `ResponseFactoryInterface`, that should outline the issue quite good.

**Comment by boesing** [comment]
> the stream factory was still outstanding for the 3.x series

That was probably my fault. I did not properly detected the callable in https://github.com/mezzio/mezzio-router/pull/23 :-/

File: src/Middleware/ImplicitHeadMiddleware.php (hunk 2 above)
Comment (gsteel) [reply src/Middleware/ImplicitHeadMiddleware.php]: I've added a decorator to wrap the callable so that we can type hint on `StreamFactoryInterface` but retained the ability to pass a callable to the constructor. 

The factory now always provides a `StreamFactoryInterface` so I can drop callable from the constructor but what about users providing custom factories? BC break or not?

File: src/Middleware/ImplicitHeadMiddlewareFactory.php (hunk 3 above)
Comment (gsteel) [reply src/Middleware/ImplicitHeadMiddlewareFactory.php]: Hopefully resolved? Preference is switched back to prefer the callable first.

File: src/Middleware/ImplicitHeadMiddleware.php [hunk 4]
```diff
@@ -44,19 +48,20 @@ class ImplicitHeadMiddleware implements MiddlewareInterface
...
 
     /**
-     * @param callable(): StreamInterface $streamFactory A factory capable of returning an empty
-     *     StreamInterface instance to inject in a response.
+     * @param (callable(): StreamInterface)|StreamFactoryInterface $streamFactory A factory capable of returning
+     *                                                                            an empty StreamInterface instance to
+     *                                                                            inject in a response.
      */
-    public function __construct(private RouterInterface $router, callable $streamFactory)
+    public function __construct(private RouterInterface $router, callable|StreamFactoryInterface $streamFactory)
```
Comment (ghostwriter) [inline src/Middleware/ImplicitHeadMiddleware.php:58]: we should make this `public function __construct(private RouterInterface $router, private StreamFactoryInterface $streamFactory)`.

If users want to use a `callable` they can pass in a `CallableDecorator` that implements `StreamFactoryInterface`.

File: src/Middleware/ImplicitOptionsMiddleware.php [hunk 5]
```diff
@@ -45,15 +45,14 @@
...
-    private $responseFactory;
+    private ResponseFactoryInterface $responseFactory;
 
     /**
      * @param (callable():ResponseInterface)|ResponseFactoryInterface $responseFactory A factory capable of returning an
      *     empty ResponseInterface instance to return for implicit OPTIONS
      *     requests.
      */
-    public function __construct($responseFactory)
+    public function __construct(callable|ResponseFactoryInterface $responseFactory)
```
Comment (ghostwriter) [inline src/Middleware/ImplicitOptionsMiddleware.php:55]: we should make this `public function __construct(private ResponseFactoryInterface $responseFactory)`

File: src/Middleware/MethodNotAllowedMiddleware.php [hunk 6]
```diff
@@ -33,13 +33,12 @@
...
 {
-    /** @var ResponseFactoryInterface */
-    private $responseFactory;
+    private ResponseFactoryInterface $responseFactory;
 
     /**
      * @param (callable():ResponseInterface)|ResponseFactoryInterface $responseFactory
      */
-    public function __construct($responseFactory)
+    public function __construct(callable|ResponseFactoryInterface $responseFactory)
```
Comment (ghostwriter) [inline src/Middleware/MethodNotAllowedMiddleware.php:41]: we should make this `public function __construct(private ResponseFactoryInterface $responseFactory)`

**Comment by ghostwriter** [comment]
review is based on the assumption that this PR will target v4.

**Comment by gsteel** [comment]
@ghostwriter 

> review is based on the assumption that this PR will target v4.

This patch is targeting 3.x - In v4, callable will be dropped along with the decorators

@boesing - when you get a chance, can you review again please?

**Comment by boesing** [comment]
Can you probably adapt the unit tests from the other PRs we've already created?
I have created some weird edge-case tests to verify we do not oversee something.

I havent had a full look on this as I currently dont want to spend time on non-SM or psalm stuff.
https://github.com/mezzio/mezzio-authorization/pull/11/files#diff-2a887fd425446ffddd77e23296b0972d2d60b1692d7151e1ade245217fc7d5d6 has some special checks regarding decorators and stuff, not quite sure if its worth to adapt these changes here or if we are good to go. 
Maybe some1 else can have a look on this or you take some more time until I've finished with SM v4. My next target would've been mezzio v4 AFAIR.

**Comment by gsteel** [comment]
Pretty confident everything is covered - the StreamFactory only has 1 responsibility here and that is to generate an empty stream for implicit HEAD requests.

Perhaps @weierophinney or @Ocramius could review to get this one over the line - we'll need it to get v4 out the door right?

**Review by weierophinney — APPROVED** [review]
Functionality looks sane, and tests cover the changes. :ship:

File: src/Stream/CallableStreamFactoryDecorator.php [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        return ($this->streamFactory)();
+    }
+
+    /** @inheritDoc */
+    public function createStreamFromFile(string $filename, string $mode = 'r'): StreamInterface
+    {
+        throw new RuntimeException('This method will not be implemented');
+    }
+
+    /** @inheritDoc */
```
Comment (Ocramius) [inline src/Stream/CallableStreamFactoryDecorator.php:38]: We should probably document `@returns never`?

File: src/Stream/CallableStreamFactoryDecorator.php [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        $this->streamFactory = $streamFactory;
+    }
+
+    /** @inheritDoc */
+    public function createStream(string $content = ''): StreamInterface
+    {
+        return ($this->streamFactory)();
+    }
+
+    /** @inheritDoc */
```
Comment (Ocramius) [inline src/Stream/CallableStreamFactoryDecorator.php:32]: We should probably document `@returns never`?

**Review by Ocramius — APPROVED** [review]
:ship: LGTM, thanks @gsteel!
